segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -120,295 +139,572 @@
+
+                // Pin it if it is already pinned.
+                transactionData.setPinned(transactionData.isPinned() || dataFromBuffer.isPinned());
 
-            // If we reach here then it means transaction is safe to commit.
-            // Step 4: Insert
-            long committedVersion = version.incrementAndGet();
-            HashMap<String, TransactionData> toAdd = new HashMap<String, TransactionData>();
-            for (String key : modifiedKeys) {
-                TransactionData data = txnData.get(key);
-                data.setVersion(committedVersion);
-                toAdd.put(key, data);
+                // Set the database object.
+                transactionData.setDbObject(dataFromBuffer.getDbObject());
+            } else {
+                Preconditions.checkState(entry.getValue().isPinned(), "Why are we here??");
             }
-            bufferedTxnData.putAll(toAdd);
         }
+    }
 
-        //  Step 5 : evict if required.
-        if (bufferedTxnData.size() > maxEntriesInTxnBuffer) {
-            bufferedTxnData.entrySet().removeIf(entry -> entry.getValue().isPersisted() && !entry.getValue().isPinned());
+    /**
+     * Evict entries if needed.
+     * Only evict keys that are persisted, not pinned or active.
+     */
+    private CompletableFuture<Void> evictIfNeeded() {
+        if (isEvictionRunning.compareAndSet(false, true)) {
+            val limit = 1 + maxEntriesInTxnBuffer / CACHE_EVICTION_RATIO;
+            if (bufferCount.get() > maxEntriesInTxnBuffer) {
+                val toEvict = bufferedTxnData.entrySet().parallelStream()
+                        .filter(entry -> entry.getValue().isPersisted() && !entry.getValue().isPinned()
+                                && !activeKeys.contains(entry.getKey()))
+                        .map(Map.Entry::getKey)
+                        .limit(limit)
+                        .collect(Collectors.toList());
+                int i = 0;
+                for (val key : toEvict) {
+                    // synchronize so that we don't accidentally delete a key that becomes active after check here.
+                    synchronized (evictionLock) {
+                        if (0 == activeKeys.count(key)) {
+                            // Synchronization prevents error when key becomes active between the check and remove.
+                            // Move the key to cache
+                            cache.put(key, bufferedTxnData.get(key));
+                            // Remove from buffer.
+                            bufferedTxnData.remove(key);
+                            i++;
+                        }
+                    }
+                }
+                bufferCount.addAndGet(-1 * i);
+            }
+            isEvictionRunning.set(false);
+            log.debug("Entries evicted from transaction buffer.");
         }
-
-        //  Step 6: finally clear
-        txnData.clear();
+        return CompletableFuture.completedFuture(null);
```

Comment:
Why do you return a CompletableFuture? Make your method return `void` if it's not async.

Reply:
fixed